### PR TITLE
[JENKINS-52325] - If copy of artifact with attributes throws error, retry without attributes

### PR DESCRIPTION
**Pull request by zbynek**
See [JENKINS-52325](https://issues.jenkins-ci.org/browse/JENKINS-52325).
Catch error when copying files. After first file fails, retry this file without COPY_ATTRIBUTES flag and copy further files also without this flag (to avoid copy - fail - rollback - retry cycles)

### Proposed changelog entries

* Entry 1: Internal: when archiving artifacts, handle the case when Jenkins can copy files but cannot change attributes like modification time (regression in TODO)

### Desired reviewers

@jvz

### CODE REVIEW

File: core/src/main/java/hudson/FilePath.java [hunk 1]
```diff
@@ -2480,13 +2480,24 @@ public Integer invoke(File base, VirtualChannel channel) throws IOException {
...
                         mkdirsE(target.getParentFile());
-                        Files.copy(fileToPath(f), fileToPath(writing(target)),
-                                StandardCopyOption.COPY_ATTRIBUTES, StandardCopyOption.REPLACE_EXISTING);
+                        Path targetPath = fileToPath(writing(target));
+                        if(!exceptionEncountered) {
+                            try {
+                                Files.copy(fileToPath(f), targetPath,
+                                    StandardCopyOption.COPY_ATTRIBUTES, StandardCopyOption.REPLACE_EXISTING);
+                            } catch (IOException e) {
+                                exceptionEncountered = true;
```
Comment (daniel-beck) [inline core/src/main/java/hudson/FilePath.java]: Great opportunity for a log message telling users that their terrible setup prevents this (or perhaps only if the alternative operation is successful for the first time).

Also, would prefer the exception not be discarded. If in doubt, log `FINE`.

File: core/src/main/java/hudson/FilePath.java [hunk 2]
```diff
@@ -2480,13 +2480,24 @@ public Integer invoke(File base, VirtualChannel channel) throws IOException {
...
+                            }
+                        }
+                        if(exceptionEncountered) {
```
[the remark this answers is not in the document]
Comment (zbynek) [reply core/src/main/java/hudson/FilePath.java]: we need to reach this branch when exception is thrown for the first time, so `else` won't work

**Review by jvz — CHANGES_REQUESTED** [review]
It might be cleaner to refactor this copy-retry for a single file into another method. Something like `static Optional<IOException> tryCopyWithAttributes(...)` maybe? And I agree with @daniel-beck about logging the bad configuration, especially if there's a link to docs or something. :)

**Comment by daniel-beck** [comment]
> It might be cleaner to refactor this copy-retry for a single file into another method.

AFAIUI this deliberately implements the optimization to only try once to copy attributes per call. Taking this into consideration I doubt it'd be much nicer that way.

**Comment by jvz** [comment]
Oh right, good point. No need to refactor that. Would just like the exception logged.

**Comment by zbynek** [comment]
I've added some log messages. Are any other changes needed?

**Review by jvz — APPROVED** [review]
Looks good to me!

File: core/src/main/java/hudson/FilePath.java [hunk 3]
```diff
@@ -2480,16 +2483,37 @@ public Integer invoke(File base, VirtualChannel channel) throws IOException {
...
                         mkdirsE(target.getParentFile());
-                        Files.copy(fileToPath(f), fileToPath(writing(target)),
-                                StandardCopyOption.COPY_ATTRIBUTES, StandardCopyOption.REPLACE_EXISTING);
+                        Path targetPath = fileToPath(writing(target));
+                        exceptionEncountered = exceptionEncountered || !tryCopyWithAttributes(f, targetPath);
+                        if (exceptionEncountered) {
+                            Files.copy(fileToPath(f), targetPath, StandardCopyOption.REPLACE_EXISTING);
+                            if (!logMessageShown) {
+                                LOGGER.log(Level.WARNING, 
+                                    "Attributes removed when copying to {0}, please make sure Jenkins process owns the directory.",
```
[the remark this answers is not in the document]
Comment (daniel-beck) [reply core/src/main/java/hudson/FilePath.java]: I don't know. It's certainly fairly spammy output in configurations that are affected. OTOH there's somewhat broken behavior as a result. Let's see what others think.

**Review by daniel-beck — APPROVED** [review]
This looks reasonable. We can always adjust log level based on user feedback.

**Review by oleg-nenashev — APPROVED** [review]
The solution looks good though having private methods in anonymous classes is quite exotic IMHO

**Comment by oleg-nenashev** [comment]
Thanks @zbynek ! It should land in the next weekly release

**Pull request merged**